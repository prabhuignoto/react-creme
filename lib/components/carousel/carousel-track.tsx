import classNames from "classnames";
import React, { useMemo } from "react";
import { CarouselButton } from "./carousel-button";
import { CarouselTrackModel } from "./carousel-model";
import "./carousel-track.scss";

const CarouselTrack: React.FunctionComponent<CarouselTrackModel> = ({
  length,
  handleSelection,
  activeIndex,
  direction,
  onNext,
  onPrevious,
  hidePrevious,
  hideNext,
  focusable,
}: CarouselTrackModel) => {
  const carouselTrackClass = useMemo(
    () => classNames(["rc-carousel-track", `rc-carousel-track-${direction}`]),
    []
  );

  const carouselContainerClass = useMemo(
    () =>
      classNames("rc-carousel-track-wrapper", {
        [`rc-carousel-track-wrapper-${direction}`]: true,
      }),
    [direction]
  );

  const focusableProps = useMemo(() => {
    return {
      tabIndex: focusable ? 0 : -1,
      onKeyDown: (event: React.KeyboardEvent) => {
        if (event.key === "ArrowRight") {
          onNext();
        } else if (event.key === "ArrowLeft") {
          onPrevious();
        }
      },
    };
  }, []);

  return (
    <div className={carouselContainerClass}>
      <CarouselButton
        onClick={onPrevious}
        position="left"
        hide={hidePrevious}
        direction={direction}
        focusable={focusable}
      />
      <ul className={carouselTrackClass} role="list">
        {Array.from({ length }).map((_, index) => (
          <li
            key={index}
            role="listitem"
            {...focusableProps}
            className={classNames([
              "rc-carousel-track-item",
              index === activeIndex ? "rc-carousel-track-item-selected" : "",
            ])}
            onClick={() => handleSelection(index)}
          ></li>
        ))}
      </ul>
      <CarouselButton
        onClick={onNext}
        position="right"
        hide={hideNext}
        direction={direction}
        focusable={focusable}
      />
    </div>
  );
};

export { CarouselTrack };
