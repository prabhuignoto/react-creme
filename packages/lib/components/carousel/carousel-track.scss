@use '../../design/core.scss';
@use '../../design/list.scss';
@use '../../design/theme.scss';

.rc-carousel-track {
  &.rc-carousel-track-horizontal {
    @extend %center;
    @extend %list-horizontal;
  }

  &.rc-carousel-track-vertical {
    @extend %list;
    width: 3rem;
  }
}

.rc-carousel-track-item {
  border: 1px solid theme.$secondary;
  border-radius: 50%;
  height: 10px;
  margin: 4px;
  width: 10px;
}

.rc-carousel-track-item:hover {
  cursor: pointer;
}

.rc-carousel-track-item-selected {
  background: theme.$primary;
  border: 1px solid theme.$primary;
}

.rc-carousel-icon {
  @extend %icon-md;
}

.rc-carousel-track-wrapper-horizontal {
  @extend %center;
}

.rc-carousel-track-wrapper-vertical {
  @extend %list;
}
