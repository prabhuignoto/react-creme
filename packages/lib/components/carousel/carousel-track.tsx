import classNames from 'classnames';
import React, { useMemo } from 'react';
import { CarouselButton } from './carousel-button';
import { CarouselTrackProps } from './carousel-model';
import styles from './carousel-track.module.scss';
import rootStyles from './carousel.module.scss';

const CarouselTrack: React.FunctionComponent<CarouselTrackProps> = ({
  length,
  handleSelection,
  activeIndex,
  direction,
  onNext,
  onPrevious,
  hidePrevious,
  hideNext,
  focusable,
}: CarouselTrackProps) => {
  const carouselTrackClass = useMemo(
    () =>
      classNames([
        styles.track,
        direction === 'horizontal'
          ? styles.track_horizontal
          : styles.track_vertical,
      ]),
    []
  );

  const carouselContainerClass = useMemo(
    () =>
      classNames([
        rootStyles.track_wrapper,
        direction === 'horizontal'
          ? styles.track_wrapper_horizontal
          : styles.track_wrapper_vertical,
        // [styles[`track-wrapper-${direction}`]]: true,
      ]),
    [direction]
  );

  return (
    <div className={carouselContainerClass}>
      <CarouselButton
        onClick={onPrevious}
        position="left"
        hide={hidePrevious}
        direction={direction}
        focusable={focusable}
        label="Previous"
      />
      <ul className={carouselTrackClass} role="list">
        {Array.from({ length }).map((_, index) => (
          <li
            key={index}
            role="listitem"
            className={classNames([
              styles.track_item,
              index === activeIndex ? styles.track_item_selected : '',
            ])}
            onClick={() => handleSelection(index)}
          ></li>
        ))}
      </ul>
      <CarouselButton
        onClick={onNext}
        position="right"
        hide={hideNext}
        direction={direction}
        focusable={focusable}
        label="Next"
      />
    </div>
  );
};

export { CarouselTrack };
