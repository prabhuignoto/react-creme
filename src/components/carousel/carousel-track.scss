@use "../../design/layout.scss";
@use "../../design/list.scss";
@use "../../design/colors.scss";
@use "../../design/icon.scss";

.rc-carousel-track {
  &.rc-carousel-track-horizontal {
    @extend %center;
    @extend %list-horizontal;
  }

  &.rc-carousel-track-vertical {
    @extend %list;
    width: 3rem;
  }
}

.rc-carousel-track-item {
  @extend %shadow-inset;
  border: 1px solid colors.$secondary;
  border-radius: 50%;
  height: 10px;
  margin: 4px;
  width: 10px;
}

.rc-carousel-track-item:hover {
  cursor: pointer;
}

.rc-carousel-track-item-selected {
  background: colors.$primary;
  border: 1px solid colors.$primary;
}

.rc-carousel-icon {
  @extend %icon-sm;
}

.rc-carousel-track-wrapper-horizontal {
  @extend %center;
}

.rc-carousel-track-wrapper-vertical {
  @extend %list;
}
