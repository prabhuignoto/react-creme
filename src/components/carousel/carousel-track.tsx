import classNames from "classnames";
import React, { useMemo } from "react";
import { CarouselButton } from "./carousel-button";
import { CarouselTrackModel } from "./carousel-model";
import "./carousel-track.scss";

const CarouselTrack: React.FunctionComponent<CarouselTrackModel> = ({
  length,
  handleSelection,
  activeIndex,
  direction,
  onNext,
  onPrevious,
  hidePrevious,
  hideNext,
}: CarouselTrackModel) => {
  const carouselTrackClass = useMemo(
    () => classNames(["rc-carousel-track", `rc-carousel-track-${direction}`]),
    []
  );

  const carouselContainerClass = useMemo(
    () =>
      classNames("rc-carousel-track-wrapper", {
        [`rc-carousel-track-wrapper-${direction}`]: true,
      }),
    [direction]
  );

  return (
    <div className={carouselContainerClass}>
      <CarouselButton
        onClick={onPrevious}
        position="left"
        hide={hidePrevious}
        direction={direction}
      />
      <ul className={carouselTrackClass} role="list">
        {Array.from({ length }).map((_, index) => (
          <li
            key={index}
            role="listitem"
            className={classNames([
              "rc-carousel-track-item",
              index === activeIndex ? "rc-carousel-track-item-selected" : "",
            ])}
            onClick={() => handleSelection(index)}
          ></li>
        ))}
      </ul>
      <CarouselButton
        onClick={onNext}
        position="right"
        hide={hideNext}
        direction={direction}
      />
    </div>
  );
};

export { CarouselTrack };
