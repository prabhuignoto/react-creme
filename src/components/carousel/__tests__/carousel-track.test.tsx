import { fireEvent, render } from "@testing-library/react";
import React from "react";
import { CarouselTrack } from "../carousel-track";

const handler = jest.fn();

describe("Carousel Track", () => {
  it("Should render the track", () => {
    const { getByRole, getAllByRole } = render(
      <CarouselTrack
        length={3}
        direction="horizontal"
        activeIndex={1}
        handleSelection={handler}
      />
    );

    expect(getByRole("list")).toHaveClass("carousel-track-horizontal");
    expect(getAllByRole("listitem")).toHaveLength(3);
  });

  it("Should call the handler", () => {
    const { getAllByRole } = render(
      <CarouselTrack
        length={3}
        direction="horizontal"
        activeIndex={1}
        handleSelection={handler}
      />
    );

    fireEvent.click(getAllByRole("listitem")[2]);

    expect(handler).toBeCalledWith(2);
  });

  it("Should the active selection is selected", () => {
    const { getAllByRole } = render(
      <CarouselTrack
        length={3}
        direction="horizontal"
        activeIndex={1}
        handleSelection={handler}
      />
    );

    expect(getAllByRole("listitem")[1]).toHaveClass(
      "carousel-track-item-selected"
    );
  });
});
