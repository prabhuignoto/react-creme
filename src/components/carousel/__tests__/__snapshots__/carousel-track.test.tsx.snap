// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel Track Should render the track snapshot 1`] = `
<ul
  class="rc-carousel-track rc-carousel-track-horizontal"
  role="list"
>
  <li
    class="rc-carousel-track-item"
    role="listitem"
    tabindex="-1"
  />
  <li
    class="rc-carousel-track-item rc-carousel-track-item-selected"
    role="listitem"
    tabindex="-1"
  />
  <li
    class="rc-carousel-track-item"
    role="listitem"
    tabindex="-1"
  />
</ul>
`;
